Hive scripts and the cilium shell

joamaki opened the pull request:
Cilium shell is a new simpler way for implementing debugging abilities for the cilium-agent based on a simple command language. The same command language can also be used to implement complex integration tests. By sharing the same language for debugging and testing it becomes easier to write complex integration tests (as the commands become familiar from debugging) and it makes sure our debugging commands are also tested. The simple way of adding commands (compared to editing openapi.yaml and modifying cilium-dbg) also significantly lowers the barrier for adding debug tooling and also as an added bonus reduces the maintenance overhead of Cilium derivatives.

See the "asciinema" links at the bottom for demos!

Commands can be added by any cell by calling cell.Provide with hive.NewScriptCmd:
```
  var Cell = cell.Module(
    ...
    cell.Provide(newScriptCommand),
  )

  func newScriptCommand(t *Thing) hive.ScriptCmdOut {
    return hive.NewScriptCmd(
      "thing",
      script.Command(
        script.CmdUsage{Summary: "Inspect the thing"},
          func(s *script.State, args ...string) (script.WaitFunc, error) {
            return func(*script.State) (stdout, stderr string, err error) {
              stdout = fmt.Sprintf("Thing status: %s\n", t.Status())
                return
              }, nil
	  },
       ),
    )
  }
```
This can be now used in "cilium-dbg shell":
```
  $ kubectl exec -it -n kube-system ds/cilium -- cilium-dbg shell
  ...
  cilium> thing
  [stdout]
  Thing status: ...

  $ kubectl exec -it -n kube-system ds/cilium -- cilium-dbg shell thing
  Thing status: ...
  $ 
```

The same command can be used in an integration test:
```
  func TestThing(t *testing.T) {
    ctx := context.Background()
    log := hivetest.Logger(t)
    h := hive.New(thing.Cell, <deps>...)
      engine := &script.Engine{
        Cmds: h.ScriptCommands(log),
      }
      require.NoError(t, h.Start(log, ctx), "Start")
      t.Cleanup(func() {
        assert.NoError(t, h.Stop(log, ctx), "Stop")
      })
      scripttest.Test(
        t,
        ctx,
        func() *script.Engine { return engine },
        []string{},
	"testdata/*.txtar",
    )
  }
```

An example test script "testdata/test.txtar" could now look something like:
```
  thing
  grep ^Thing status: OK$
```

More information about this can be found from:
- Blog post on test scripts (which inspired me to do all of this): https://atlasgo.io/blog/2024/09/09/how-go-tests-go-test 
- https://github.com/cilium/hive/pull/17
- https://github.com/cilium/statedb/pull/57
- Reconciler test case: https://github.com/cilium/statedb/blob/main/reconciler/testdata/incremental.txtar
- Complicated integration test example: https://github.com/joamaki/cilium/blob/pr/joamaki/hivescript-load-balancing/pkg/loadbalancer/experimental/testdata/nodeport.txtar
- Shell demo: https://asciinema.org/a/qwsPRRT1GT4GfIAaby6o1ZyJ7
- Adding a new shell command: https://asciinema.org/a/xIUd2dVeOvojFrxIptpKPnqWa
- Integration testing demo: https://asciinema.org/a/2Oqd9yWxahmerI6PuMPAmGh9m

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,31 @@
+
+func shell(cmd *cobra.Command, args []string) {
+	conn, err := net.Dial("unix", "/tmp/ciliumshell.sock")
+	if err != nil {
```
we might want to retry before panic.
will this only panic the cilium-dbg?

joamaki committed:
daemon,cilium-dbg: Cilium shell

Cilium shell is a new simpler way for implementing debugging abilities
for the cilium-agent based on a simple script language. The same scripting
language can also be used to implement complex integration tests. By sharing
the same language for debugging and testing it becomes easier to write complex
integration tests (as the commands become familiar from debugging) and it makes
sure our debugging commands are actually tested.

Commands can be added by any cell by calling cell.Provide with hive.NewScriptCmd:

  var Cell = cell.Module(
    ...
    cell.Provide(newScriptCommand),
  )

  func newScriptCommand(t *Thing) hive.ScriptCmdOut {
    return hive.NewScriptCmd(
		"thing",
		script.Command(
			script.CmdUsage{Summary: "Inspect the thing"},
			func(s *script.State, args ...string) (script.WaitFunc, error) {
			  return func(*script.State) (stdout, stderr string, err error) {
			    stdout = fmt.Sprintf("Thing status: %s\n", t.Status())
				return
 			  }, nil
			},
		),
	)
  }

This can be now used in "cilium-dbg shell":

  $ kubectl exec -it -n kube-system ds/cilium -- cilium-dbg shell
  ...
  cilium> thing
  [stdout]
  Thing status: ...

The same command can be used in an integration test:

  func TestThing(t *testing.T) {
    ctx := context.Background()
    log := hivetest.Logger(t)
    h := hive.New(thing.Cell, <deps>...)
	engine := &script.Engine{
      Cmds: h.ScriptCommands(log),
	}
	require.NoError(t, h.Start(log, ctx), "Start")
	t.Cleanup(func() {
		assert.NoError(t, h.Stop(log, ctx), "Stop")
	})
    scripttest.Test(
	  t,
	  context.Background(),
	  func() *script.Engine { return engine },
	  []string{},
	  "testdata/*.txtar",
	)
  }

An example test script "testdata/test.txtar" count now look something like:

  thing
  grep '^Thing status: OK$'

Signed-off-by: Jussi Maki <jussi@isovalent.com>

learnitall reviewed (approved):
Great idea! I'm excited about this. I read through the blog post and the rest of the links that you added and I'm on board with the vision. I have a couple of questions for you, but they may be more appropriate for discussion in slack:

* One of the core concepts that was mentioned in the blog post was the idea that these tests are able to be sandboxed and can be executed in parallel. Is this a goal for use with Cilium? If so, how will tests that rely on datapath state be implemented?
* What is the migration plan for existing tests, if any? Will this framework be expected as the primary method for testing hive cells, while other integration test frameworks that Cilium currently uses will be utilized for non-modularized code?
* IIUC the framework is not reliant on a kubernetes cluster, therefore coverage for integration with kubernetes or other external APIs is missing. Is this a future use case that the framework will handle? Or will this type of testing continue to be handled by existing integration testing code?

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+		case "[stdout]":
+		default:
+			w.Write(line)
+			w.Write(newline)
```
should we handle the error from Write()?

joamaki replied:
We don't care as much if writing to stdout fails, but since this might be used with something else I'll check the errors here.

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+
+func shell(cmd *cobra.Command, args []string) {
+	if len(args) > 0 {
+		shellExchange(os.Stdout, "%s", strings.Join(args, " "))
```
shellExchange() returns an error but we are ignoring it here

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+
+}
+
+func shell(cmd *cobra.Command, args []string) {
```
nit: 
```suggestion
func shell(_ *cobra.Command, args []string) {
```

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+	// Try to set the terminal to raw mode (so that cursor keys work etc.)
+	restore, err := script.MakeRaw(0)
+	if err != nil {
+		panic(err)
```
better return the error from here and handle in shell()?

joamaki replied:
Since it's non-fatal I'll just print the error to stderr and continue. Effect is that cursor keys etc. won't work.

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+	defer signal.Stop(sigs)
+	signal.Notify(sigs, os.Interrupt)
+
+	term := term.NewTerminal(stdReadWriter, "cilium> ")
```
nit: collides with imported package `golang.org/x/term`

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+		// is meant for interactive use we'll try to be helpful and retry the dialing until
+		// agent comes up.
+		conn, err := dialShell()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+			os.Exit(1)
+		}
```
should we handle the errors in shell()?

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+		for {
+			line, err := term.ReadLine()
+			if err != nil {
+				return
```
return the error?

joamaki replied:
I think we can just assume that failing to read from the terminal is the same as EOF and just stop.

ovidiutirla commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,179 @@
+			for {
+				line, _, err := bio.ReadLine()
+				if err != nil {
+					break outer
```
can we return and defer close(stop) here instead of using the break block?

joamaki replied:
huh weird, why didn't I write it that way in the first place...

joamaki replied:
doh, because we don't want to return, but we want to redial the connection instead when the user presses ctrl-c (to interrupt a command). I refactored this to at least avoid the break label.

ovidiutirla commented on daemon/cmd/shell.go:
```diff
@@ -0,0 +1,133 @@
+	var lc net.ListenConfig
+	l, err := lc.Listen(ctx, "unix", defaults.ShellSockPath)
+	if err != nil {
+		return fmt.Errorf("Listen(%s): %w", defaults.ShellSockPath, err)
```
```suggestion
		return fmt.Errorf("listen(%s): %w", defaults.ShellSockPath, err)
```

ovidiutirla commented on daemon/cmd/shell.go:
```diff
@@ -0,0 +1,133 @@
+	for ctx.Err() == nil {
+		conn, err := l.Accept()
+		if err != nil {
+			return fmt.Errorf("Accept(%s): %w", defaults.ShellSockPath, err)
```
```suggestion
			return fmt.Errorf("accept(%s): %w", defaults.ShellSockPath, err)
```

ovidiutirla reviewed (approved):
Looking forward for features to integrate this.

tommyp1ckles commented on cilium-dbg/cmd/shell.go:
```diff
@@ -0,0 +1,194 @@
+		if err == nil {
+			break
+		}
+		if time.Now().After(until) {
```
Does this mean that we still wait for the net.Dial to return, then check if it exceeded the timeout afterwards? Shouldn't we just use `net.DialContext` with a timeout ctx?

joamaki replied:
Ah might as well. The retries/timeout was really just to retry for convenience when someone does "cilium-dbg shell" before the agent is fully started up (which happens pretty often).

EDIT: Ah but for non-interactive "cilium-dbg shell <cmd>" use having a DialContext with timeout makes sense.

pippolo84 reviewed (approved):
LGTM! :rocket:
